File: backend/app/schemas/leadership_task.py
"""Pydantic Schemas cho Nhiem vu & Y kien chi dao cua Ban Chi huy Lu doan (v7.7.0)."""

from datetime import date, datetime
from typing import Literal, Optional
from pydantic import BaseModel, ConfigDict, Field

CommanderRole = Literal["lu_truong", "chinh_uy", "lu_pho_tmt", "lu_pho_hckt", "pho_chinh_uy"]
TargetBranch = Literal["tham_muu", "chinh_tri", "hau_can_ky_thuat", "toan_lu_doan"]
Urgency = Literal["hoa_toc", "khan", "thuong"]
# Vong doi trang thai: dang_thuc_hien -> (submit_report) da_bao_cao
#   -> (review_task) da_hoan_thanh | can_bo_sung ; can_bo_sung + submit_report -> da_bao_cao
TaskStatus = Literal["dang_thuc_hien", "da_bao_cao", "da_hoan_thanh", "can_bo_sung"]
ReviewStatus = Literal["da_hoan_thanh", "can_bo_sung"]

COMMANDER_ROLE_LABELS = {
    "lu_truong": "Lữ đoàn trưởng (Chỉ đạo chung mọi mặt về Chính quyền)",
    "chinh_uy": "Chính uỷ Lữ đoàn (Chỉ đạo chung mọi mặt bên Đảng)",
    "lu_pho_tmt": "Phó Lữ đoàn trưởng kiêm TMT (Chỉ đạo chuyên ngành Tham mưu - Tác chiến - TTLL)",
    "lu_pho_hckt": "Phó Lữ đoàn trưởng HC-KT (Chỉ đạo chuyên ngành Hậu cần - Kỹ thuật - VKTB)",
    "pho_chinh_uy": "Phó Chính uỷ Lữ đoàn (Điều hành các Tổ chức Quần chúng & Tổ chức Đảng)",
}

URGENCY_LABELS = {
    "hoa_toc": "Hoả tốc",
    "khan": "Khẩn",
    "thuong": "Thường",
}

TASK_STATUS_LABELS = {
    "dang_thuc_hien": "Đang thực hiện",
    "da_bao_cao": "Đã báo cáo, chờ bút phê",
    "da_hoan_thanh": "Đã hoàn thành",
    "can_bo_sung": "Cần báo cáo bổ sung",
}


class LeadershipTaskCreate(BaseModel):
    commander_role: CommanderRole = Field(..., description="lu_truong | chinh_uy | lu_pho_tmt | lu_pho_hckt | pho_chinh_uy")
    title: str = Field(..., min_length=3, max_length=255)
    content: str = Field(..., min_length=5)
    target_branch: TargetBranch = Field("toan_lu_doan", description="tham_muu | chinh_tri | hau_can_ky_thuat | toan_lu_doan")
    assigned_unit_id: Optional[int] = None
    urgency: Urgency = Field("thuong", description="hoa_toc | khan | thuong")
    deadline: Optional[date] = None


class LeadershipTaskReport(BaseModel):
    report_content: str = Field(..., min_length=5, description="Báo cáo tiến độ và kết quả thực hiện chỉ đạo")


class LeadershipTaskReview(BaseModel):
    status: ReviewStatus = Field("da_hoan_thanh", description="Kết luận bút phê: da_hoan_thanh | can_bo_sung")
    review_note: str = Field(..., min_length=2, description="Ý kiến đánh giá, nhận xét của Chỉ huy Lữ đoàn")


class LeadershipTaskOut(BaseModel):
    id: int
    commander_role: str
    commander_role_label: str
    commander_id: int
    commander_name: str
    title: str
    content: str
    target_branch: str
    target_branch_label: str
    assigned_unit_id: Optional[int] = None
    assigned_unit_name: Optional[str] = None
    urgency: str
    urgency_label: str
    deadline: Optional[date] = None
    status: str
    status_label: str
    report_content: Optional[str] = None
    reported_by_id: Optional[int] = None
    reported_by_name: Optional[str] = None
    reported_at: Optional[datetime] = None
    review_note: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LeadershipTaskListResponse(BaseModel):
    items: list[LeadershipTaskOut]
    total: int

File: frontend/src/types/leadershipTask.ts
/**
 * Types cho Bàn làm việc Chỉ đạo Ban Chỉ huy Lữ đoàn.
 * Đồng bộ openapi v8.0.0 (2026-09-10) — chuẩn hoá mô hình trạng thái + siết enum.
 */

export type CommanderRole =
  | 'lu_truong'
  | 'chinh_uy'
  | 'lu_pho_tmt'
  | 'lu_pho_hckt'
  | 'pho_chinh_uy';

export type CommanderBranch =
  | 'tham_muu'
  | 'chinh_tri'
  | 'hau_can_ky_thuat'
  | 'toan_lu_doan';

export type TaskUrgency = 'thuong' | 'khan' | 'hoa_toc';

/**
 * Trạng thái chỉ đạo (v8.0.0 — 4 giá trị). Vòng đời:
 *   dang_thuc_hien → (đơn vị nộp báo cáo) da_bao_cao
 *   da_bao_cao → (BCH bút phê) da_hoan_thanh | can_bo_sung
 *   can_bo_sung → (nộp lại) da_bao_cao
 */
export type TaskStatus =
  | 'dang_thuc_hien'
  | 'da_bao_cao'
  | 'da_hoan_thanh'
  | 'can_bo_sung';

/** Kết luận bút phê — body POST /leadership-tasks/{id}/review (v8.0.0). */
export type LeadershipReviewStatus = 'da_hoan_thanh' | 'can_bo_sung';

export interface LeadershipTask {
  id: number;
  commander_role: CommanderRole;
  commander_role_label?: string;
  commander_id: number;
  commander_name: string;
  title: string;
  content: string;
  target_branch: CommanderBranch;
  target_branch_label?: string;
  assigned_unit_id?: number | null;
  assigned_unit_name?: string | null;
  urgency: TaskUrgency;
  urgency_label?: string;
  deadline?: string | null;
  status: TaskStatus;
  status_label?: string;
  report_content?: string | null;
  reported_by_id?: number | null;
  reported_by_name?: string | null;
  reported_at?: string | null;
  review_note?: string | null;
  reviewed_at?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface LeadershipTaskCreatePayload {
  commander_role: CommanderRole;
  title: string;
  content: string;
  target_branch?: CommanderBranch;
  assigned_unit_id?: number | null;
  urgency?: TaskUrgency;
  deadline?: string | null;
}

export interface LeadershipTaskReportPayload {
  /** Tối thiểu 5 ký tự (openapi v8.0.0). */
  report_content: string;
}

export interface LeadershipTaskReviewPayload {
  /** Mặc định backend: da_hoan_thanh (openapi v8.0.0). */
  status?: LeadershipReviewStatus;
  /** Bắt buộc, tối thiểu 2 ký tự (openapi v8.0.0). */
  review_note: string;
}

export interface LeadershipTaskListResponse {
  items: LeadershipTask[];
  total: number;
}

import type { IconName } from '../components/Icon'

export const COMMANDER_ROLE_META: Record<
  CommanderRole,
  { label: string; title: string; desc: string; icon: IconName; badgeClass: string }
> = {
  lu_truong: {
    label: 'Lữ đoàn trưởng',
    title: 'Đồng chí Lữ đoàn trưởng',
    desc: 'Chỉ đạo chung mọi mặt về Chính quyền (Quân sự · Tác chiến · Kế hoạch SSCĐ)',
    icon: 'star',
    badgeClass: 'bg-red-700 text-amber-300 border-amber-400/30',
  },
  chinh_uy: {
    label: 'Chính uỷ Lữ đoàn',
    title: 'Đồng chí Chính uỷ Lữ đoàn',
    desc: 'Chỉ đạo chung mọi mặt bên Đảng (Đảng uỷ · Cấp uỷ · CTĐ - CTCT · Cán bộ)',
    icon: 'shield',
    badgeClass: 'bg-rose-800 text-rose-100 border-rose-500/30',
  },
  lu_pho_tmt: {
    label: 'Phó Lữ trưởng kiêm TMT',
    title: 'Phó Lữ đoàn trưởng kiêm Tham mưu trưởng',
    desc: 'Chỉ đạo chuyên ngành Tham mưu · Tác chiến · Huấn luyện · TTLL SSCĐ (d1, d2)',
    icon: 'layers',
    badgeClass: 'bg-blue-800 text-blue-100 border-blue-400/30',
  },
  lu_pho_hckt: {
    label: 'Phó Lữ trưởng HC-KT',
    title: 'Phó Lữ đoàn trưởng Hậu cần – Kỹ thuật',
    desc: 'Chỉ đạo chuyên ngành Hậu cần – Kỹ thuật · VKTB · Khí tài TTLL · Xe máy (c5, Trạm)',
    icon: 'clipboard',
    badgeClass: 'bg-amber-800 text-amber-100 border-amber-400/30',
  },
  pho_chinh_uy: {
    label: 'Phó Chính uỷ',
    title: 'Đồng chí Phó Chính uỷ Lữ đoàn',
    desc: 'Điều hành các Tổ chức Quần chúng (Đoàn TN, Hội PN, Công đoàn) & Tổ chức Đảng',
    icon: 'users',
    badgeClass: 'bg-emerald-800 text-emerald-100 border-emerald-400/30',
  },
};

export const TASK_URGENCY_META: Record<
  TaskUrgency,
  { label: string; badgeClass: string }
> = {
  hoa_toc: {
    label: 'Hoả tốc',
    badgeClass: 'bg-red-600 text-white animate-pulse border-red-300',
  },
  khan: {
    label: 'Khẩn',
    badgeClass: 'bg-amber-500 text-slate-900 font-bold border-amber-300',
  },
  thuong: {
    label: 'Thường',
    badgeClass: 'bg-slate-700 text-slate-200 border-slate-600',
  },
};

export const TASK_STATUS_META: Record<
  TaskStatus,
  { label: string; badgeClass: string; icon: IconName }
> = {
  dang_thuc_hien: {
    label: 'Đang triển khai',
    badgeClass: 'bg-blue-500/20 text-blue-300 border-blue-500/40',
    icon: 'layers',
  },
  da_bao_cao: {
    label: 'Đã báo cáo, chờ bút phê',
    badgeClass: 'bg-amber-500/20 text-amber-300 border-amber-500/40',
    icon: 'file',
  },
  da_hoan_thanh: {
    label: 'Đã hoàn thành',
    badgeClass: 'bg-emerald-500/20 text-emerald-300 border-emerald-500/40',
    icon: 'check',
  },
  can_bo_sung: {
    label: 'Cần bổ sung',
    badgeClass: 'bg-rose-500/20 text-rose-300 border-rose-500/40',
    icon: 'alert-triangle',
  },
};
